src/Microsoft.ML.Api/DataViewConstructionUtils.cs

```diff
@@ -129,112 +129,192 @@ private Delegate CreateGetter(int index)
+                        }
+                        else if (outputType.GetElementType() == typeof(short))
+                        {
+                            Ch.Assert(colType.ItemType == NumberType.I2);
+                            return CreateArrayGetterDelegate<short, DvInt2>(index, x => x);
+                        }
+                        else if (outputType.GetElementType() == typeof(short?))
+                        {
+                            Ch.Assert(colType.ItemType == NumberType.I2);
+                            return CreateArrayGetterDelegate<short?, DvInt2>(index, x => x ?? DvInt2.NA);
+                        }
+                        else if (outputType.GetElementType() == typeof(sbyte))
+                        {
+                            Ch.Assert(colType.ItemType == NumberType.I1);
+                            return CreateArrayGetterDelegate<sbyte, DvInt1>(index, x => x);
                         }
+                        else if (outputType.GetElementType() == typeof(sbyte?))
+                        {
+                            Ch.Assert(colType.ItemType == NumberType.I1);
+                            return CreateArrayGetterDelegate<sbyte?, DvInt1>(index, x => x ?? DvInt1.NA);
+                        }
+                        else if (outputType.GetElementType() == typeof(bool))
+                        {
+                            Ch.Assert(colType.ItemType.IsBool);
+                            return CreateArrayGetterDelegate<bool, DvBool>(index, x => x);
+                        }
+                        else if (outputType.GetElementType() == typeof(bool?))
+                        {
+                            Ch.Assert(colType.ItemType.IsBool);
+                            return CreateArrayGetterDelegate<bool?, DvBool>(index, x => x ?? DvBool.NA);
+                        }
+
                         // T[] -> VBuffer<T>
-                        Ch.Assert(outputType.GetElementType() == colType.ItemType.RawType);
-                        del = CreateArrayToVBufferGetter<int>;
+                        if (outputType.GetElementType().IsGenericType && outputType.GetElementType().GetGenericTypeDefinition() == typeof(Nullable<>))
+                            Ch.Assert(Nullable.GetUnderlyingType(outputType.GetElementType()) == colType.ItemType.RawType);
+                        else
+                            Ch.Assert(outputType.GetElementType() == colType.ItemType.RawType);
+                        del = CreateDirectArrayGetterDelegate<int>;
+                        genericType = outputType.GetElementType();
                     }
                     else if (colType.IsVector)
                     {
                         // VBuffer<T> -> VBuffer<T>
                         // REVIEW: Do we care about accomodating VBuffer<string> -> VBuffer<DvText>?
+                        // REVIEW: why it's int and not long?
```

Comment:
>// REVIEW: why it's int and not long? [](start = 24, length = 37)

This is a strange place to add this comment. 
We are doing MarshalInvoke here, right. We have the 'del' as a generic on some token type (it was historically always int), then in line 283 we make the correct getter.